rpcutils/rpcclient.go

```diff
@@ -631,3 +633,185 @@ func APITransaction(client txhelpers.VerboseTransactionGetter, txid *chainhash.H
 
 	return
 }
+
+// TxInput returns the input information for a tx at the given input index.
+func TxInput(client txhelpers.VerboseTransactionGetter, txid string, inputIndex uint32) (*chainjson.Vin, error) {
+	txhash, err := chainhash.NewHashFromStr(txid)
+	if err != nil {
+		return nil, fmt.Errorf("invalid transaction id %s", txid)
+	}
+
+	txraw, err := GetTransactionVerboseByID(client, txhash)
+	if err != nil {
+		return nil, fmt.Errorf("GetTransactionVerboseByID failed for %v: %v", txid, err)
+	}
+
+	if int(inputIndex) >= len(txraw.Vin) {
+		return nil, fmt.Errorf("tx does not contain input at index %d", inputIndex)
+	}
+
+	return &txraw.Vin[inputIndex], nil
+}
+
+// SpendingTxsFn is any function that returns spending information for the
+// outputs of a transaction.
+type SpendingTxsFn func(string) ([]string, []uint32, []uint32, error)
+
+// TxAtomicSwapsInfo checks the outputs of the specified transaction for possible
+// atomic swap contracts and the inputs for possible swap redemptions or refunds.
+// Returns all contracts, redemptions and refunds that were found.
+func TxAtomicSwapsInfo(client txhelpers.VerboseTransactionGetter, params *chaincfg.Params, txid string,
+	spendingTxsFn SpendingTxsFn) (*apitypes.TxAtomicSwaps, error) {
+
+	txhash, err := chainhash.NewHashFromStr(txid)
+	if err != nil {
+		return nil, fmt.Errorf("invalid transaction id %s", txid)
+	}
+
+	txraw, err := GetTransactionVerboseByID(client, txhash)
+	if err != nil {
+		return nil, fmt.Errorf("GetTransactionVerboseByID failed for %v: %v", txid, err)
+	}
+
+	tx := &apitypes.TxAtomicSwaps{
+		TxID:        txid,
+		Contracts:   make(map[uint32]*apitypes.AtomicSwap),
+		Redemptions: make(map[uint32]*apitypes.AtomicSwap),
+		Refunds:     make(map[uint32]*apitypes.AtomicSwap),
+	}
+
+	// Check if tx is stake tree tx or coinbase tx and return empty swap info.
+	for _, input := range txraw.Vin {
```

Comment:
Moving this check before `tx`'s initialization would save some allocations since you'd return nil if the tx happens to be a coinbase or a stakebase.